pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -627,17 +627,110 @@ describe('StablePhantomPool', () => {
+      });
+
+      it('fails if no user data', async () => {
+        await expect(pool.join({ data: '0x' })).to.be.revertedWith('Transaction reverted without a reason');
+      });
+
+      it('fails if wrong user data', async () => {
+        const wrongUserData = ethers.utils.defaultAbiCoder.encode(['address'], [lp.address]);
+
+        await expect(pool.join({ data: wrongUserData })).to.be.revertedWith('Transaction reverted without a reason');
+      });
+
+      describe('join exact tokens in for BPT out', () => {
+        context('not in recovery mode', () => {
+          itJoinsGivenExactTokensInCorrectly();
         });
+
+        context('in recovery mode', () => {
+          sharedBeforeEach('enable recovery mode', async () => {
+            await pool.enableRecoveryMode(admin);
+          });
+
+          itJoinsGivenExactTokensInCorrectly();
+        });
+
+        function itJoinsGivenExactTokensInCorrectly() {
+          it('fails if not initialized', async () => {
+            await expect(pool.joinGivenIn({ recipient, amountsIn: initialBalances })).to.be.revertedWith(
+              'UNINITIALIZED'
+            );
+          });
+
+          context('once initialized', () => {
+            let expectedBptOut: BigNumberish;
+            let amountsIn: BigNumberish[];
+
+            sharedBeforeEach('initialize pool', async () => {
+              await pool.init({ recipient, initialBalances });
+              bptIndex = await pool.getBptIndex();
+              amountsIn = ZEROS.map((n, i) => (i != bptIndex ? fp(0.1) : n));
+
+              expectedBptOut = await pool.estimateBptOut(
+                await pool.upscale(amountsIn),
+                await pool.upscale(initialBalances)
+              );
+            });
+
+            it('grants BPT for exact tokens', async () => {
+              const previousBptBalance = await pool.balanceOf(recipient);
+              const minimumBptOut = pct(expectedBptOut, 0.99);
+
+              const result = await pool.joinGivenIn({ amountsIn, minimumBptOut, recipient, from: recipient });
+
+              // Amounts in should be the same as initial ones
+              expect(result.amountsIn).to.deep.equal(amountsIn);
+
+              // Make sure received BPT is closed to what we expect
+              const currentBptBalance = await pool.balanceOf(recipient);
+              expect(currentBptBalance.sub(previousBptBalance)).to.be.equalWithError(expectedBptOut, 0.0001);
+            });
+
+            async function joinGivenIn(): Promise<void> {
+              const minimumBptOut = pct(expectedBptOut, 0.99);
+
+              const result = await pool.queryJoinGivenIn({ amountsIn, minimumBptOut });
+
+              expect(result.amountsIn).to.deep.equal(amountsIn);
+              expect(result.bptOut).to.be.equalWithError(expectedBptOut, 0.0001);
+            }
+
+            it('can tell how much BPT it will give in return', async () => {
+              await joinGivenIn();
+            });
```

Comment:
Seems like we only use this function here. Seems like we can inline it?

Reply:
Yes; this must have been called from multiple places at one time (maybe having to do with recovery mode)